### DISCUSSION TITLE
TGN Example

### DISCUSSION THREAD

**Discussion by gao462** [pyg-team/pytorch_geometric#3469] [Q&A]
The example of `examples/tgn.py` seems to only work for events without temporal node features (or maybe I misunderstand), but the original TGN paper supports the temporal node features.

Can I ask for a better explanation or documentation on how to put temporal node features into TGNMemory? Currently, I force to merge source and destination node features in the msg input (thus msg_dim is `2 * node_dim + edge_dim`).

**Comment by rusty1s** [comment]
As far as I see, you can directly add them as inputs to `gnn` as well, see [here](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/tgn.py#L121), indexed based on `n_id` from a global node feature matrix store. Does that work in your case?

**Reply by gao462** [reply]
Thank you for the reply. For me, it is a valid way to add node features.

But I am indeed more interested in the usage of `TGNMemory`. In the original paper, the equation (2) describes how to embed a node event as a message vector (which requires a node message network different from edge message network), but I do not see any parameter that can create a node message network.
I know that one way to do this is has MLPs to sync the node and edge feature size, then feed the node event as a self-loop edge event, but it may conflict with real self-loop event (although not exists in my current usage). So I am wondering is there any arguments to create an independent node event message network?

**Comment by gao462** [comment]
I think directly adding node features is not the best match for me.
The TGNMemory will give the latest memory state of each node, however, in a batch, we may have different temporal node features for the same node (same node at different times have different features), thus the shape of node memory states and raw node temporal features can not fit.
I checked the TGN paper, it seems that they use only the latest raw node feature in a batch (which I think is not correct since their attn convolution design is claimed to be based on TGAT which use all temporal node features). I also checked their repo, and their implementation even doesn't support using raw node features as they described in the paper.
I think the TGN model just does not support raw node features well in their design.

**Reply by rusty1s** [reply]
It's been some time since i looked into TGN. Do you know which steps are necessary to incorporate node features as part of the memory as described in the TGN paper?

**Reply by gao462** [reply]
In current TGN implementation, `update_state(self, src, dst, t, raw_msg)` need to become `update_state_edge(self, src, dst, t, raw_msg)`, and we also need a new function `update_state_node(self, nids, t, raw_feat)` (same thing also needed for `__update_msg_store__(...)`). Furthermore, `TGNMemory` will need a new message module `self.msg_node_module` other than `self.msg_s_module`, `self.msg_d_module` since the edge message and node message input size can be different.

**Reply by rusty1s** [reply]
@emalgorithm
